src/Disks/IO/CachedOnDiskReadBufferFromFile.cpp

```diff
@@ -950,39 +1046,115 @@ bool CachedReadBufferFromRemoteFS::nextImplStep()
 
     if (size == 0 && file_offset_of_buffer_end < read_until_position)
     {
-        std::optional<size_t> cache_file_size;
-        if (auto * cache_file_reader = dynamic_cast<ReadBufferFromFileDescriptor *>(implementation_buffer.get()))
-            cache_file_size = cache_file_reader->getFileSize();
+        size_t cache_file_size = getFileSizeFromReadBuffer(*implementation_buffer);
+        auto cache_file_path = getFileNameFromReadBuffer(*implementation_buffer);
 
         throw Exception(
             ErrorCodes::LOGICAL_ERROR,
-            "Having zero bytes, but range is not finished: file offset: {}, reading until: {}, read type: {}, cache file size: {}",
+            "Having zero bytes, but range is not finished: file offset: {}, starting offset: {}, "
+            "reading until: {}, read type: {}, cache file size: {}, cache file path: {}, "
+            "cache file offset: {}, current file segment: {}",
             file_offset_of_buffer_end,
+            first_offset,
             read_until_position,
             toString(read_type),
-            cache_file_size ? std::to_string(*cache_file_size) : "None");
+            cache_file_size ? std::to_string(cache_file_size) : "None",
+            cache_file_path,
+            implementation_buffer->getFileOffsetOfBufferEnd(),
+            file_segment->getInfoForLog());
     }
 
     return result;
 }
 
-off_t CachedReadBufferFromRemoteFS::seek(off_t offset, int whence)
+off_t CachedOnDiskReadBufferFromFile::seek(off_t offset, int whence)
 {
-    if (initialized)
-        throw Exception(ErrorCodes::CANNOT_SEEK_THROUGH_FILE, "Seek is allowed only before first read attempt from the buffer");
+    if (initialized && !allow_seeks)
+    {
+        throw Exception(
+            ErrorCodes::CANNOT_SEEK_THROUGH_FILE,
+            "Seek is allowed only before first read attempt from the buffer");
+    }
 
-    if (whence != SEEK_SET)
+    size_t new_pos = offset;
+
+    if (allow_seeks)
```

Comment:
> Because AFAICS SEEK_SET is allowed.

It is allowed only before first read attempt, e.g. only once. So probably need something like `allow_seeks_after_first_read`.